Libraries/Components/MaskedView/MaskedViewIOS.ios.js

```diff
@@ -0,0 +1,154 @@
+const findNodeHandle = require('ReactNative').findNodeHandle;
+const requireNativeComponent = require('requireNativeComponent');
+const ReactPropTypes = React.PropTypes;
+
+import type { StyleObj } from 'StyleSheetTypes';
+
+/**
+ * Renders the child view with a mask specified in the `renderMask` prop.
+ *
+ * ```
+ * import React from 'react';
+ * import { MaskedView, Text, View } from 'react-native';
+ *
+ * class MyMaskedView extends React.Component {
+ *   render() {
+ *     return (
+ *       <MaskedView
+ *         style={{ flex: 1 }}
+ *         renderMask={() =>
+ *           <View style={styles.maskContainerStyle}>
+ *             <Text style={styles.maskTextStyle}>
+ *               Basic Mask
+ *             </Text>
+ *           </View>
+ *         }
+ *       >
+ *         <View style={{ flex: 1, backgroundColor: 'blue' }} />
+ *       </MaskedView>
+ *     );
+ *   }
+ * }
+ * ```
+ *
+ * The above example will render a view with a blue background that fills its
+ * parent, and then mask that view with text that says "Basic Mask".
+ *
+ * The alpha channel of the view rendered by the `renderMask` prop determines how
+ * much of the view’s content and background shows through. Fully or partially
+ * opaque pixels allow the underlying content to show through but fully
+ * transparent pixels block that content.
+ *
+ */
+class MaskedViewIOS extends React.Component {
+  props: {
+    style?: StyleObj,
+    /**
+     * `MaskedView` only accepts a single child as a view to be masked.
+     */
+    children: ReactElement<*>,
+    /**
+     * Should return a React element to be rendered and applied as the
+     * mask for the child element.
+     */
+    renderMask: () => React.Element<*>,
+  };
+
+  state: {
+    maskViewNodeHandle: ?mixed,
+  } = {
+    maskViewNodeHandle: null,
+  };
+
+  static propTypes = {
+    ...ViewPropTypes,
+    renderMask: ReactPropTypes.func.isRequired,
+  };
+
+  _hasWarnedInvalidChildren = false;
+  _hasWarnedInvalidRenderMask = false;
+
+  render() {
+    if (!this.props.children || React.Children.count(this.props.children) > 1) {
+      if (!this._hasWarnedInvalidChildren) {
+        console.warn(
+          'MaskedView: MaskedView requires a single child React Element.'
+        );
+        this._hasWarnedInvalidChildren = true;
+      }
+      return null;
+    }
+
+    const target = React.Children.only(this.props.children);
+
+    if (typeof this.props.renderMask !== 'function') {
+      if (!this._hasWarnedInvalidRenderMask) {
+        console.warn(
+          'MaskedView: Invalid `renderMask` prop was passed to MaskedView. ' +
+            'Expected a function. No mask will render.'
+        );
+        this._hasWarnedInvalidRenderMask = true;
+      }
+      return <View style={this.props.style}>{target}</View>;
+    }
+
+    const maskElement = this.props.renderMask();
+    let maskElementWithRef = null;
+    if (maskElement) {
+      maskElementWithRef = React.cloneElement(maskElement, {
```

Comment:
Use `cloneReferencedElement` so that if the owner of the mask set a ref, we don't clobber it.